src/orbits/keplerian.jl

```diff
@@ -0,0 +1,532 @@
+
+    if (isnothing(ecc) || iszero(ecc)) && !isnothing(duration)
+        isnothing(R_star) && (R_star = no_units ? 1.0 : 1.0u"Rsun")
+        aR_star = compute_aor(duration, period, b, RpRs=RpRs)
+        a = R_star * aR_star
+        duration = nothing
+    end
+
+    if !isnothing(a) && isnothing(M_planet)
+        M_planet = no_units ? 0.0 : 0.0u"Msun"
+    end
+
+    if !isnothing(period) && isnothing(M_planet)
+        M_planet = no_units ? 0.0 : 0.0u"Msun"
+    end
+
+    # Compute implied stellar density
+    implied_rho_star = false
+    if all( (!isnothing).((a, period)) )
+        if any( (!isnothing).((rho_star, M_star)) )
+            throw(ArgumentError(
+                "If both `a` and `P` are given, `ρₛ` or `Mₛ` cannot be defined"
+            ))
+        end
+
+        # Default to Rₛ = 1 R⊙ if not provided
+        isnothing(R_star) && (R_star = oneunit(a))
+
+        # Compute implied mass
+        M_tot = 4.0 * π^2 * a^3 / (G * period^2)
+
+        # Compute implied density
+        M_star = M_tot - M_planet
+        rho_star = M_star / ((4.0/3.0) * π * R_star^3)
+        implied_rho_star = true
+    end
+
+    # Check combination of stellar params are valid
+    if all( isnothing.((R_star, M_star)) )
+        R_star = no_units ? 1.0 : 1.0u"Rsun"
+        isnothing(rho_star) && (M_star = oneunit(M_planet))
+    end
+
+    if !implied_rho_star && sum(isnothing.((rho_star, R_star, M_star))) ≠ 1
+        throw(ArgumentError(
+            "Must provide exactly two of: `ρₛ`, `Rₛ`, or `Mₛ` if ρₛ not implied"
+        ))
+    end
+
+    # Compute stellar params
+    if isnothing(rho_star)
+        rho_star = 3.0 * M_star / (4.0 * π * R_star^3)
+    elseif isnothing(R_star)
+        R_star = ( 3.0 * M_star / (4.0 * π * rho_star) )^(1/3)
+    else
+        M_star = 4.0 * π * R_star^3 * rho_star / 3.0
+    end
+
+    # Compute planet params
+    M_tot = M_star + M_planet
+    if isnothing(a)
+        if isnothing(aR_star)
+            a = ( G * M_tot * period^2 / (4.0 * π^2) )^(1/3)
+            aR_star = a / R_star
+        else
+            a = aR_star * R_star
+        end
+    else
+        period = 2.0 * π * a^(3/2) / (√(G * M_tot))
+        aR_star = a / R_star
+    end
+
+    return a, aR_star, period, rho_star, R_star, M_star, M_planet, duration
+end
+
+stringify_units(value::Unitful.AbstractQuantity, unit) = value
+stringify_units(value, unit) = "$value $unit"
+function Base.show(io::IO, ::MIME"text/plain", orbit::KeplerianOrbit)
+    if orbit.period isa Real
+        if isnothing(orbit.rho_planet)
+            rho_planet = nothing
+        else
+            rho_planet = orbit.rho_planet * MsunRsun_to_gcc
+        end
+        rho_star = orbit.rho_star * MsunRsun_to_gcc
+    else
+        if isnothing(orbit.rho_planet)
+            rho_planet = nothing
+        else
+            println("we here")
+            rho_planet = orbit.rho_planet |> u"g/cm^3"
+        end
+        rho_star = orbit.rho_star |> u"g/cm^3"
+    end
```

Comment:
looks like this is a bit of a debugging adventure, so I'll assume it'll get simplified like the rest of the printing code below it 😛